Laravel branch

Zavodsk0y opened the pull request:
Для токенов используется пакет Sanctum. Пока тыкал впервые, по сути набрал лишнего - jetstream к примеру.
Пробовал сначала сделать аутентификацию на laravel/passport, но потом решил отказаться от этого.
Токены шифруются дополнительно - юзеру при реге и логине выкидывает plainTextToken, который он и должен использовать.
Чутка описал в редмишке существующие эндпоинты (API.md), но только их успешное применение
Для доступа была создана отдельная модель + таблица, куда вносился идентификатор пользователя и файла + тип(автор, соавтор), ид - внешние ключи соответственно.
Напрягает валидация, потому ее придется выносить скорее всего куда-нибудь
+?при ошибках валидации и прочего, возможно лучше будет выкидывать исключения к примеру, а не респонс?
Насчет хранения также до конца непонятно, там ли я храню файлы (они же по идее в паблике должны быть вообще, а не в app/...).
Повторяющиеся респонсы - тоже большой напряг, их, как мне кажется тоже нужно будет выносить

Zavodsk0y committed:
Создание laravel-проекта

Zavodsk0y committed:
Редактируем шаблон welcome, начинаем подстраивать jetstream для работы.
composer.lock косякс

Zavodsk0y committed:
Создаем апи контроллер для аутентификации, куда включим метод на регистрацию для присвоения токена, далее туда пихнем авторизацию.
Было принято решение отказаться в пакета "Passport" из-за отказа работы. (-2 часа, санктум - спасибо).
В конфиг auth.php добавляем в гварды апи, перестаем хешировать токен
В Kernel.php добавили старт сессий, ибо без этого выдавало ошибку

Zavodsk0y committed:
Доделываем функционал гостя + начинаем прорабатывать функционал авторизованного юзера (логаут).
До конца не разобрался, как именно достать токен, принадлежащий пользователю при логине. Пришлось воспользоваться костылем и генерить для него новый токен, и уже через логаут удалять все существующие у него токены

Zavodsk0y committed:
Пытаемся в файлы, тестовая реализация. Пока пользуемся небольшими костылями для чуть более корректного вывода. Меняем конфиг nginx загружаемых файлов до 10 мб, чтобы мы могли увидеть ошибку нашего валидатора. Создаем модель + контроллер + роут. Обновляем существуюую модель

Zavodsk0y committed:
Первая реализация удаления файла. Изменили метод загрузки файла (storeAs, чтобы сохранить его в директории по айдишнику, по которому мы потом удялем из Storage).

Zavodsk0y committed:
Апдейтим наш контроллер на загрузку и удаление.
Прописываем требования нового имени для файла в случае, если такое уже есть ( (1), (2), и т.д.).
Удаляем теперь из хранилища по имени файла, а не айдишнику, ибо теперь мы сохраняем по имени

Zavodsk0y committed:
Реализуем методы удаления + скачивания файла

Zavodsk0y committed:
Первичная реализация управления доступом. Сделали для корректной проверки php.ini, в котором изменили размер макс. загружаемых файлов до 20Мб, также теперь при загрузке файлов мы добавляем запись в новую таблицу, указывая тип "автор".
Делаем новую модель + контроллер, в котором прописываем логику добавления пользователя
Считаю, что в будущем похожие-типовые респонсы нужно будет куда-то вынести, чтобы разгрузить код в контроллерах

Zavodsk0y committed:
Глобально допиливаем существующий функционал и переписываем существующий для корректной работы.
Меняем метод unauthenticated у посредника Authenticate, чтобы выдавал нам 403 статус по требованиям
Для доступа к файлам организовали новую модель, в которой храним fk ид файла, ид юзера и тип (автор, соавтор).
Сделали так, чтобы в бд сохранялись файлы с одинаковыми именами, но для разных юзеров
В хранилище файлы храним по айди+расширению, ибо другого варианта я пока не предполагаю

Zavodsk0y committed:
Меняем дефолтный адрес, группируем роуты

Zavodsk0y committed:
Чуть редачим контроллер файлов
Закидываем php.ini + провайдеры для корректной работы

Zavodsk0y committed:
Чуть редачим контроллер файлов
Закидываем php.ini + провайдеры для корректной работы

Zavodsk0y committed:
Update ApiAuthController.php

Zavodsk0y committed:
Update FileAccessController.php

Zavodsk0y committed:
Update FileController.php

Zavodsk0y committed:
Create API.md

Zavodsk0y committed:
Delete cloud/app/Http/Middleware/API.md

Zavodsk0y committed:
Create API.md

Zavodsk0y committed:
Update README.md

Zavodsk0y committed:
Update README.md

Zavodsk0y committed:
Добавляем сидер для создания пользователей

RetmixX commented on cloud/routes/api.php:
```diff
@@ -0,0 +1,37 @@
+
+Route::middleware(['auth:sanctum'])->group(function () {
+    Route::get('/logout', [ApiAuthController::class, 'logout']);
+    Route::post("/files", [FileController::class, 'uploadFiles']);
```
https://laravel.com/docs/10.x/controllers#resource-controllers

RetmixX commented on cloud/database/factories/UserFactory.php:
```diff
@@ -0,0 +1,69 @@
+            'name' => $this->faker->name(),
+            'email' => $this->faker->unique()->safeEmail(),
+            'email_verified_at' => now(),
+            'password' => '$2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi', // password
```
Проще Hash::make('password')

RetmixX commented on cloud/app/Http/Controllers/API/ApiAuthController.php:
```diff
@@ -0,0 +1,108 @@
+                'first_name' => 'required|min:2',
+                'last_name' => 'required',
+            ]);
+        } catch (ValidationException $e) {
```
https://laravel.com/docs/10.x/validation#form-request-validation

RetmixX commented on cloud/app/Http/Controllers/API/FileAccessController.php:
```diff
@@ -0,0 +1,143 @@
+
+class FileAccessController extends Controller
+{
+    public function addAccessToFile($fileId, Request $request): JsonResponse
```
В ларавеле есть автоматическое внедрение моделей
https://laravel.com/docs/10.x/routing#route-model-binding

RetmixX reviewed (changes requested):
Поправить замечания, ВЫНЕСТИ ЛОГИКУ ИЗ КОНТРОЛЛЕРОВ, В ДРУГОЕ МЕСТО

jurapro commented on cloud/app/Http/Controllers/API/ApiAuthController.php:
```diff
@@ -0,0 +1,108 @@
+        } catch (ValidationException $e) {
+            $errors = $e->errors();
+
+            return response()->json([
```
для типовых ответов лучше https://laravel.com/docs/10.x/eloquent-resources

jurapro commented on cloud/app/Http/Controllers/API/ApiAuthController.php:
```diff
@@ -0,0 +1,108 @@
+                'email' => 'required|email',
+                'password' => 'required'
+            ]);
+        } catch (ValidationException $e) {
```
в данном случае нет смысла try catch использовать, просто смотришь успешная ли валидация. вообще можно использовать исключения, но нужно их отлавливать не в самом контроллере, а в handler, переопределив стандартные классы исключения тут реализацию можно подсмотреть https://github.com/jurapro/api-share

jurapro commented on cloud/app/Http/Controllers/API/ApiAuthController.php:
```diff
@@ -0,0 +1,108 @@
+    public function logout(Request $request): JsonResponse
+    {
+
+        $user = $request->user();
```
можно через Auth()

jurapro commented on cloud/app/Http/Controllers/API/FileAccessController.php:
```diff
@@ -0,0 +1,143 @@
+            ], 403);
+        }
+
+        $userAccess = User::where('email', $email)->first();
```
Доступы делаем через политики https://laravel.com/docs/10.x/authorization#writing-policies

jurapro commented on cloud/app/Http/Controllers/API/FileAccessController.php:
```diff
@@ -0,0 +1,143 @@
+        $fileAccesses = FileAccess::where('file_id', $fileId)->get();
+
+        $usersWithAccess = [];
+        foreach ($fileAccesses as $access) {
```
не понятно для чего, видимо значит нужно рефачить

jurapro commented on cloud/app/Http/Controllers/API/FileController.php:
```diff
@@ -0,0 +1,312 @@
+        $files = $request->file('files');
+        $user = auth()->user();
+
+        if ($files) {
```
Большая вложенность, инвертируй условный оператор

jurapro replied:
Блок рефакторинга

jurapro commented on cloud/app/Http/Controllers/API/FileController.php:
```diff
@@ -0,0 +1,312 @@
+                $validator = Validator::make(array('file' => $uploadFile), $rules);
+
+                if ($validator->fails()) {
+                    $response = [
```
Кидаешь исключение или ответ и не нужно дальше else

jurapro commented on cloud/app/Http/Controllers/API/FileController.php:
```diff
@@ -0,0 +1,312 @@
+                        $counter = 1;
+                        $newFileName = $uploadFileName;
+
+                        while (true) {
```
ойойой

jurapro reviewed (changes requested):
- [ ] все замечание, которые ты сам написал справедливы. нужно выносить валидации, нужно выносить логику из контоллеров, нужно использовать исключения
- [ ] почитай замечания, писали только в начале, далее однотипно
- [ ] много грязи в коде, нужно рефачить, очень толстые методы в контроллерах, не всегда код очевиден
Попробуй отрефачить

Zavodsk0y committed:
Созданием форм-реквестов для логина, регистрации и загрузки файлов.
Выносим логику из контроллеров в раздел "Сервисы"
С внедрением модели пока не разобрался + нужно будет сделать ресурс-контроллер для файлов

Zavodsk0y committed:
Обновление контроллера на ресурсный, с учетом нюансов (диск по адресу files/disk, потому индекс же не получится использовать, так?)

Нужно ли делать также для доступа? Ибо там дестрой идет по идентификатору какому-то все равно, а у нас это передается в теле, по чему именно мы удаляем(почта)

Баг-фикс в сервисе, сейв нужно было вынести выше, чтобы видел всех пользователей, которые получили доступ к файлу, включая нового

Zavodsk0y committed:
Фиксы
Сервисы.
Добавляем исключения, которые начинаем выкидывать
Создали файл-ресурс, который уже возвращаем в методе processFile нашего сервиса
Обновили FilesUploadRequest.php (доработать, не понятно, как сделать без лишнего своего валидатора)
Обновили Handler, добавив в него исключения и проверки
Чуть-чуть подкорректировали контроллер аутентификации

RetmixX commented on cloud/app/Exceptions/Handler.php:
```diff
@@ -38,4 +42,30 @@ public function register()
             //
         });
     }
+    public function render($request, Throwable $e)
```
Попробуй такой вариант
![image](https://github.com/Zavodsk0y/laravel/assets/98283896/aa78c878-5a05-4f5a-b482-548ca930ab57)

RetmixX commented on cloud/app/Http/Services/FileService.php:
```diff
@@ -0,0 +1,238 @@
+            'name' => $uploadFileName,
+        ]);
+
+        $fileAccess = new FileAccess([
```
Можно короче Model::create([*fields*]), метод сразу создает и сохраняет в бд

RetmixX commented on cloud/app/Http/Services/FileAccessService.php:
```diff
@@ -0,0 +1,127 @@
+        $usersWithAccess = [];
+        foreach ($fileAccesses as $access) {
+            $usersWithAccess[] = [
+                'fullname' => $access->user->first_name . ' ' . $access->user->last_name,
```
"{$access->user->first_name} {$access->user->last_name}"

jurapro commented on cloud/app/Http/Controllers/API/FileAccessController.php:
```diff
@@ -47,96 +35,25 @@ public function addAccessToFile($fileId, Request $request): JsonResponse
-            ];
+        if (array_key_exists('data', $response)) {
+            return response()->json($response['data'], $response['code']);
+        } else {
```
else не нужен

jurapro commented on cloud/app/Http/Controllers/API/FileController.php:
```diff
@@ -3,107 +3,33 @@
 
+    protected $fileService;
+
+    public function __construct(FileService $fileService)
```
можно объявлять гораздо элегантнее https://sergeymukhin.com/blog/php-8-prodvizenie-svoistv-konstruktora

jurapro commented on cloud/app/Http/Services/FileAccessService.php:
```diff
@@ -0,0 +1,127 @@
+    {
+        $user = auth()->user();
+
+        if ($file->user_id !== $user->id) {
```
Доступы не тут должны быть, а в политиках

jurapro commented on cloud/app/Http/Services/FileAccessService.php:
```diff
@@ -0,0 +1,127 @@
+    {
+        $user = auth()->user();
+
+        if (!$file) {
```
Такие вещи исключения нужно кидать, а не просто респонс

jurapro commented on cloud/app/Http/Services/FileService.php:
```diff
@@ -0,0 +1,265 @@
+    {
+        $validationError = $this->validateFile($uploadFile);
+
+        if ($validationError) {
```
Валидации в сервисах не хорошо

jurapro reviewed (changes requested):
Облегчил контроллеры - это хорошо
Почти всю логику в сервисы запихал - это плохо

Валидации, проверки на доступы должны быть отдельно от логики приложения
Опционально - в нужных местах юзать исключения

Zavodsk0y committed:
Работа с политикой.
Валидатор свой выкинул в отдельный класс и директорию Validators.
Файл теперь получаем в контроллере, кидаем его в сервис, для удобного взаимодействия с политикой
Объявление сервиса переработал
Handler не перерабатывал, ибо после установки пакета Spaties он не обнаружил у меня нужный класс ошибки.
Теперь с файлами делаем File::create, вместо new File([fields]), чтобы сразу создать файл и закинуть их в бд
Кидаем исключения

jurapro commented on cloud/app/Http/Validators/FileValidator.php:
```diff
@@ -0,0 +1,18 @@
+
+use Illuminate\Support\Facades\Validator;
+
+class FileValidator
```
Наверное не так выразился, имел в видду формреквесты для валидации https://laravel.com/docs/10.x/validation#form-request-validation

Zavodsk0y replied:
Я так и не допер просто, как через формреквест обработать каждый файл и сделать так, чтобы загрузка остальных продолжалась. Сейчас же получается, если вкинуть просто формреквест, то если валидация завалится на первом файле - то остальные уже будут проигнорированы

jurapro reviewed (approved):
Все ок, кроме валидации. Нужно формреквесты юзать, свой велосипед не нужно)
